src/daft-functions/src/uri/upload.rs

```diff
@@ -28,59 +29,155 @@ impl ScalarUDF for UploadFunction {
+/// This function also parses the given folder path and trims '/' from the end of the path.
+fn instantiate_and_trim_path(
+    folder_path: &str,
+    is_single_folder: bool,
+    instantiated_folder_paths: &mut HashSet<String>,
+) -> DaftResult<String> {
+    // HACK: Creates folders if running locally. This is a bit of a hack to do it here because we'd rather delegate this to
+    // the appropriate source. However, most sources such as the object stores don't have the concept of "folders".
+    let (source, folder_path) = daft_io::parse_url(folder_path)?;
+    if matches!(source, SourceType::File) {
+        let local_prefixless_folder_path = match folder_path.strip_prefix("file://") {
+            Some(p) => p,
+            None => folder_path.as_ref(),
+        };
+        if is_single_folder {
+            // If we were given a single folder, create a directory at the given path.
+            if instantiated_folder_paths.insert(local_prefixless_folder_path.to_string()) {
+                std::fs::create_dir_all(local_prefixless_folder_path).map_err(|e| {
+                    daft_io::Error::UnableToCreateDir {
+                        path: local_prefixless_folder_path.to_string(),
+                        source: e,
+                    }
+                })?;
+            }
+        } else {
+            // If we were given row-specific paths, then we create directories at the parents of the given paths.
+            let path = Path::new(local_prefixless_folder_path);
+            if let Some(parent_dir) = path.parent() {
+                if let Some(parent_dir) = parent_dir.to_str() {
+                    if instantiated_folder_paths.insert(parent_dir.to_string()) {
+                        std::fs::create_dir_all(parent_dir).map_err(|e| {
+                            daft_io::Error::UnableToCreateDir {
+                                path: parent_dir.to_string(),
+                                source: e,
+                            }
+                        })?;
+                    }
+                }
+            }
+        }
+    }
+    Ok(folder_path.trim_end_matches('/').to_string())
+}
+
+/// Helper function that takes an utf8 array of folder paths that may have either 1 or `len` paths,
+/// and returns a Vec of `len` folder paths. This function will also instantiate and trim the given
+/// folder paths as needed.
+///
+/// If `is_single_folder` is set, the prepared path is repeated `len` times.
+/// Otherwise, we return the array of prepared paths.
+fn prepare_folder_paths(
+    arr: &Utf8Array,
+    len: usize,
+    is_single_folder: bool,
+) -> DaftResult<Vec<String>> {
+    let mut instantiated_folder_paths = HashSet::new();
+    if is_single_folder {
+        let folder_path = arr.get(0).unwrap();
+        let folder_path =
+            instantiate_and_trim_path(folder_path, true, &mut instantiated_folder_paths)?;
+        Ok(repeat(folder_path).take(len).collect())
+    } else {
+        assert_eq!(arr.len(), len);
```

Comment:
Could we use a `debug_assert_eq!` here?

Reply:
Yeah can do (tbf I don't believe that this check will be expensive relative to the actual upload which should dominate runtime)